diff --git a/server/token.c b/server/token.c
index 7deddc46645..65714df0f60 100644
--- a/server/token.c
+++ b/server/token.c
@@ -144,6 +144,7 @@ static const struct object_ops token_ops =
     no_add_queue,              /* add_queue */
     NULL,                      /* remove_queue */
     NULL,                      /* signaled */
+    NULL,                      /* get_esync_fd */
     NULL,                      /* satisfied */
     no_signal,                 /* signal */
     no_get_fd,                 /* get_fd */
