diff --git a/server/thread.h b/server/thread.h
index e4332df4abd..773d67c7147 100644
--- a/server/thread.h
+++ b/server/thread.h
@@ -89,6 +89,8 @@ struct thread
     timeout_t              creation_time; /* Thread creation time */
     timeout_t              exit_time;     /* Thread exit time */
     struct token          *token;         /* security token associated with this thread */
+    int                    esync_fd;      /* esync file descriptor (signalled on exit) */
+    int                    esync_apc_fd;  /* esync apc fd (signalled when APCs are present) */
 };
 
 struct thread_snapshot
