diff --git a/server/registry.c b/server/registry.c
index 964e2ebd58f..2d5583ffaf0 100644
--- a/server/registry.c
+++ b/server/registry.c
@@ -159,6 +159,7 @@ static const struct object_ops key_ops =
     no_add_queue,            /* add_queue */
     NULL,                    /* remove_queue */
     NULL,                    /* signaled */
+    NULL,                    /* get_esync_fd */
     NULL,                    /* satisfied */
     no_signal,               /* signal */
     no_get_fd,               /* get_fd */
