diff --git a/server/process.h b/server/process.h
index 313c36ab241..b36a5cea328 100644
--- a/server/process.h
+++ b/server/process.h
@@ -95,6 +95,7 @@ struct process
     struct list          rawinput_devices;/* list of registered rawinput devices */
     const struct rawinput_device *rawinput_mouse; /* rawinput mouse device, if any */
     const struct rawinput_device *rawinput_kbd;   /* rawinput keyboard device, if any */
+    int                  esync_fd;        /* esync file descriptor (signaled on exit) */
 };
 
 struct process_snapshot
