diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 98c3f2d9508..f7f385f332e 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -44,6 +44,7 @@
 #include "wine/server.h"
 #include "ntdll_misc.h"
 #include "ddk/wdm.h"
+#include "esync.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(module);
 WINE_DECLARE_DEBUG_CHANNEL(relay);
@@ -4246,6 +4247,8 @@ void __wine_process_init(void)
     peb->ProcessHeap = RtlCreateHeap( HEAP_GROWABLE, NULL, 0, 0, NULL, NULL );
     peb->LoaderLock = &loader_section;
 
+    esync_init();
+
     init_directories();
     init_user_process_params( info_size );
 
