diff --git a/server/request.c b/server/request.c
index 200c2697dbd..03054a07c18 100644
--- a/server/request.c
+++ b/server/request.c
@@ -96,6 +96,7 @@ static const struct object_ops master_socket_ops =
     no_add_queue,                  /* add_queue */
     NULL,                          /* remove_queue */
     NULL,                          /* signaled */
+    NULL,                          /* get_esync_fd */
     NULL,                          /* satisfied */
     no_signal,                     /* signal */
     no_get_fd,                     /* get_fd */
