diff --git a/dlls/ntdll/om.c b/dlls/ntdll/om.c
index 0b562513489..6c74ce535d1 100644
--- a/dlls/ntdll/om.c
+++ b/dlls/ntdll/om.c
@@ -37,6 +37,7 @@
 #include "windef.h"
 #include "winternl.h"
 #include "ntdll_misc.h"
+#include "esync.h"
 #include "wine/server.h"
 #include "wine/exception.h"
 
@@ -390,6 +391,9 @@ NTSTATUS close_handle( HANDLE handle )
     NTSTATUS ret;
     int fd = server_remove_fd_from_cache( handle );
 
+    if (do_esync())
+        esync_close( handle );
+
     SERVER_START_REQ( close_handle )
     {
         req->handle = wine_server_obj_handle( handle );
